### PR TITLE
Use `slices.Sort()` instead of `sort.Strings()` or `sort.Ints()`

### PR DESCRIPTION
**Pull request by charleskorn**
#### What this PR does

This PR was sparked by https://github.com/grafana/mimir/pull/3639#discussion_r1042880752. It replaces all usage of `sort.Strings()` with [`slices.Sort()`](https://pkg.go.dev/golang.org/x/exp@v0.0.0-20221211140036-ad323defaf05/slices#Sort), and adds a linting rule to enforce this in the future.

I'm opening this as a draft for discussion because:
* this is a wide-ranging change
* `golang.org/x/exp/slices` is still considered an experimental package (although Prometheus is now using it [widely](https://github.com/prometheus/prometheus/pull/11318))
* I haven't benchmarked these changes (but everywhere else that has replaced `sort.Strings()` with `slices.Sort()` has seen substantial improvements, see below)

---

##### Performance

https://github.com/grafana/mimir/pull/3639#discussion_r1042880752 and https://github.com/prometheus/prometheus/pull/11054 show substantial performance improvements when using `slices.Sort()` over `sort.Strings()`. This improved performance is because `slices.Sort()` does not require the overhead of an interface call for every comparison. (If we were using Go 1.18 or earlier, we'd also benefit from the improved pdquicksort sorting algorithm in `slices`, but this has been implemented in the `sort` package as of Go 1.19.)

Given the pattern of significant performance improvements in a number of scenarios, I haven't benchmarked the changes in this PR. If we do want to take a more measured approach, I'm very open to that feedback.

##### Future work

We could also replace `sort.Float64s()` with `slices.Sort()`, however, this requires more careful consideration as [the documentation for `slices.Sort()`](https://pkg.go.dev/golang.org/x/exp@v0.0.0-20221211140036-ad323defaf05/slices#Sort) warns that it may not handle NaN values correctly.

#### Which issue(s) this PR fixes or relates to

(none)

#### Checklist

- [n/a] Tests updated
- [n/a] Documentation added
- [n/a] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by bboreham** [comment]
> This improved performance is because slices.Sort() does not require the overhead of an interface call for every comparison.

This is partially true; as I said in the Prometheus PR.

> Some of the speedup comes from [not] an `interface .Less()` call; some comes from `exp/slices` using "pattern-defeating quicksort(pdqsort)" which is much better on already-sorted data.

**Review by bboreham — COMMENTED** [review]
> I haven't benchmarked the changes in this PR

I don't want to encourage this stance.  Measuring is always better than assuming.
A quick search of the source code suggests there might not be an existing one covering any of these changes, which in itself is interesting. We should be interested in the performance of `LabelNames`, for instance.

I perhaps wouldn't have changed the tests; they aren't very performance-sensitive and it can be nice to have some diversity of algorithm between implementation and test.  But then your lint rule wouldn't work.

**Comment by charleskorn** [comment]
> This is partially true; as I said in the Prometheus PR.
> 
> > Some of the speedup comes from [not] an `interface .Less()` call; some comes from `exp/slices` using "pattern-defeating quicksort(pdqsort)" which is much better on already-sorted data.

I think this changed with the Go 1.19 release - [the release notes](https://go.dev/doc/go1.19) mention changing the algorithm for the `sort` package to pdqsort, and `sort.Strings()` calls into [an implementation of pdqsort](https://cs.opensource.google/go/go/+/refs/tags/go1.19.4:src/sort/zsortinterface.go;drc=888047c31007c0ad96764254fa8afb1417ddfd40;l=61) which looks almost identical to [the one in `golang.org/x/exp/slices`](https://cs.opensource.google/go/x/exp/+/ad323def:slices/zsortordered.go;drc=732eee02a75a571a8347fdf60af8c8b9e3ca9254;l=63).

File: Makefile [hunk 1]
```diff
@@ -323,6 +323,15 @@ lint: check-makefiles
 		github.com/prometheus/client_golang/prometheus.{NewCounter,NewCounterVec,NewCounterFunc,NewGauge,NewGaugeVec,NewGaugeFunc,NewSummary,NewSummaryVec,NewHistogram,NewHistogramVec}=github.com/prometheus/client_golang/prometheus/promauto.With" \
 		./pkg/...
 
+	# Use the faster slices.Sort where we can.
+	# See https://github.com/prometheus/prometheus/pull/11054 and https://github.com/prometheus/prometheus/pull/11318 for more
+	# background information on slices.Sort() and the real-world performance impact it can have.
```
[the remark this answers is not in the document]
Comment (charleskorn) [reply Makefile]: Fair enough - I've removed that in 874341d.

**Comment by pracucci** [comment]
@charleskorn I would keep a very practical approach, benchmarking few of the changes in the hot path, but not all of them. If we don't find regressions on few of them, I would just move on and merge.

I think this PR touches somewhat hot paths in the following functions. Do we have any benchmark touching any of those?

```
Distributor.LabelValuesForLabelName
Distributor.LabelNames
mergeExemplarQueryResponses()
matrixMerge()
fetchLabelValuesFromStore()
blockLabelNames()
BinaryReader.LabelValues()
BinaryReader.LabelNames()
PostingOffsetTableV1.LabelValues()
```

**Comment by charleskorn** [comment]
tl;dr: for the four instances I benchmarked, we see improvements of anywhere from 2.2% to 16%, and none regressed. I'm happy to keep benchmarking the other changes, but the results are universally positive so far.

---

These are some results from benchmarks that we already had or were easily able to be changed to cover some of the affected methods:

<details>
<summary><code>BinaryReader.LabelNames()</code></summary>

```
name                                                 old time/op    new time/op    delta
LabelNames/20Names100Values/BinaryReader-10            1.69µs ± 2%    1.48µs ± 2%  -12.21%  (p=0.008 n=5+5)
LabelNames/20Names500Values/BinaryReader-10            1.69µs ± 1%    1.49µs ± 2%  -11.98%  (p=0.008 n=5+5)
LabelNames/20Names1000Values/BinaryReader-10           1.70µs ± 2%    1.49µs ± 1%  -12.25%  (p=0.008 n=5+5)
LabelNames/50Names100Values/BinaryReader-10            4.40µs ± 1%    3.73µs ± 1%  -15.06%  (p=0.008 n=5+5)
LabelNames/50Names500Values/BinaryReader-10            4.46µs ± 5%    3.73µs ± 1%  -16.45%  (p=0.008 n=5+5)
LabelNames/50Names1000Values/BinaryReader-10           4.37µs ± 1%    3.72µs ± 1%  -14.75%  (p=0.008 n=5+5)
LabelNames/100Names100Values/BinaryReader-10           9.51µs ± 0%    8.18µs ± 0%  -14.00%  (p=0.008 n=5+5)
LabelNames/100Names500Values/BinaryReader-10           9.53µs ± 1%    8.19µs ± 1%  -14.05%  (p=0.008 n=5+5)
LabelNames/100Names1000Values/BinaryReader-10          9.56µs ± 1%    8.18µs ± 1%  -14.45%  (p=0.008 n=5+5)
LabelNames/200Names100Values/BinaryReader-10           20.6µs ± 0%    17.5µs ± 0%  -15.00%  (p=0.008 n=5+5)
LabelNames/200Names500Values/BinaryReader-10           20.6µs ± 0%    17.5µs ± 0%  -15.09%  (p=0.008 n=5+5)
LabelNames/200Names1000Values/BinaryReader-10          20.8µs ± 3%    17.4µs ± 1%  -16.08%  (p=0.008 n=5+5)

name                                                 old alloc/op   new alloc/op   delta
LabelNames/20Names100Values/BinaryReader-10              424B ± 0%      400B ± 0%   -5.66%  (p=0.008 n=5+5)
LabelNames/20Names500Values/BinaryReader-10              424B ± 0%      400B ± 0%   -5.66%  (p=0.008 n=5+5)
LabelNames/20Names1000Values/BinaryReader-10             424B ± 0%      400B ± 0%   -5.66%  (p=0.008 n=5+5)
LabelNames/50Names100Values/BinaryReader-10              968B ± 0%      944B ± 0%   -2.48%  (p=0.008 n=5+5)
LabelNames/50Names500Values/BinaryReader-10              968B ± 0%      944B ± 0%   -2.48%  (p=0.008 n=5+5)
LabelNames/50Names1000Values/BinaryReader-10             968B ± 0%      944B ± 0%   -2.48%  (p=0.008 n=5+5)
LabelNames/100Names100Values/BinaryReader-10           1.86kB ± 0%    1.84kB ± 0%   -1.29%  (p=0.008 n=5+5)
LabelNames/100Names500Values/BinaryReader-10           1.86kB ± 0%    1.84kB ± 0%   -1.29%  (p=0.008 n=5+5)
LabelNames/100Names1000Values/BinaryReader-10          1.86kB ± 0%    1.84kB ± 0%   -1.29%  (p=0.008 n=5+5)
LabelNames/200Names100Values/BinaryReader-10           3.53kB ± 0%    3.50kB ± 0%   -0.68%  (p=0.008 n=5+5)
LabelNames/200Names500Values/BinaryReader-10           3.53kB ± 0%    3.50kB ± 0%   -0.68%  (p=0.008 n=5+5)
LabelNames/200Names1000Values/BinaryReader-10          3.53kB ± 0%    3.50kB ± 0%   -0.68%  (p=0.008 n=5+5)

name                                                 old allocs/op  new allocs/op  delta
LabelNames/20Names100Values/BinaryReader-10              4.00 ± 0%      3.00 ± 0%  -25.00%  (p=0.008 n=5+5)
LabelNames/20Names500Values/BinaryReader-10              4.00 ± 0%      3.00 ± 0%  -25.00%  (p=0.008 n=5+5)
LabelNames/20Names1000Values/BinaryReader-10             4.00 ± 0%      3.00 ± 0%  -25.00%  (p=0.008 n=5+5)
LabelNames/50Names100Values/BinaryReader-10              4.00 ± 0%      3.00 ± 0%  -25.00%  (p=0.008 n=5+5)
LabelNames/50Names500Values/BinaryReader-10              4.00 ± 0%      3.00 ± 0%  -25.00%  (p=0.008 n=5+5)
LabelNames/50Names1000Values/BinaryReader-10             4.00 ± 0%      3.00 ± 0%  -25.00%  (p=0.008 n=5+5)
LabelNames/100Names100Values/BinaryReader-10             4.00 ± 0%      3.00 ± 0%  -25.00%  (p=0.008 n=5+5)
LabelNames/100Names500Values/BinaryReader-10             4.00 ± 0%      3.00 ± 0%  -25.00%  (p=0.008 n=5+5)
LabelNames/100Names1000Values/BinaryReader-10            4.00 ± 0%      3.00 ± 0%  -25.00%  (p=0.008 n=5+5)
LabelNames/200Names100Values/BinaryReader-10             4.00 ± 0%      3.00 ± 0%  -25.00%  (p=0.008 n=5+5)
LabelNames/200Names500Values/BinaryReader-10             4.00 ± 0%      3.00 ± 0%  -25.00%  (p=0.008 n=5+5)
LabelNames/200Names1000Values/BinaryReader-10            4.00 ± 0%      3.00 ± 0%  -25.00%  (p=0.008 n=5+5)
```

</details>

<details>
<summary><code>BinaryReader.LabelValues()</code> and <code>PostingOffsetTableV1.LabelValues()</code></summary>

Note: `StreamBinaryReader` uses `PostingOffsetTableV1` - I've tested it through `StreamBinaryReader` here to save me constructing a slightly different benchmark setup just for it.

```
name                                      old time/op    new time/op    delta
LabelValuesIndexV1/StreamBinaryReader-10    4.19µs ± 1%    3.55µs ± 2%  -15.23%  (p=0.008 n=5+5)
LabelValuesIndexV1/BinaryReader-10          4.18µs ± 1%    3.53µs ± 0%  -15.62%  (p=0.008 n=5+5)

name                                      old alloc/op   new alloc/op   delta
LabelValuesIndexV1/StreamBinaryReader-10      960B ± 0%      936B ± 0%   -2.50%  (p=0.008 n=5+5)
LabelValuesIndexV1/BinaryReader-10            960B ± 0%      936B ± 0%   -2.50%  (p=0.008 n=5+5)

name                                      old allocs/op  new allocs/op  delta
LabelValuesIndexV1/StreamBinaryReader-10      3.00 ± 0%      2.00 ± 0%  -33.33%  (p=0.008 n=5+5)
LabelValuesIndexV1/BinaryReader-10            3.00 ± 0%      2.00 ± 0%  -33.33%  (p=0.008 n=5+5)
```

</details>

<details>
<summary><code>mergeExemplarQueryResponses</code></summary>

```
name               old time/op    new time/op    delta
MergeExemplars-10    1.15ms ± 0%    1.13ms ± 0%  -2.20%  (p=0.008 n=5+5)

name               old alloc/op   new alloc/op   delta
MergeExemplars-10    1.28MB ± 0%    1.28MB ± 0%    ~     (p=0.794 n=5+5)

name               old allocs/op  new allocs/op  delta
MergeExemplars-10     5.06k ± 0%     5.06k ± 0%  -0.02%  (p=0.008 n=5+5)
```

</details>

**Review by pracucci — APPROVED** [review]
Thanks @charleskorn for the benchmarks. I'm happy to move forward and merge it. Can you rebase it, please?

**Review by pracucci — APPROVED** [review]
LGTM, thanks!
